src/nvim/ex_cmds.c

```diff
@@ -4960,3 +4961,27 @@ void ex_oldfiles(exarg_T *eap)
     }
   }
 }
+
+void ex_trust(exarg_T *eap)
+{
+  const char *const p = skiptowhite(eap->arg);
+  char *arg1 = xmemdupz(eap->arg, (size_t)(p - eap->arg));
+  const char *action = "allow";
+  const char *path = eap->arg;
+
+  if (strcmp(arg1, "++deny") == 0) {
+    path = skipwhite(p);
+    action = "deny";
+  } else if (strcmp(arg1, "++remove") == 0) {
+    path = skipwhite(p);
+    action = "remove";
+  }
+
+  if (path[0] == '\0') {
+    path = NULL;
+  }
+
+  nlua_trust(action, path);
+
+  xfree(arg1);
+}
```

Comment:
More robust argument parsing, `++deny` and `++remove` are mutually exclusive, and it handles invalid or duplicate `++options`.

Needs the `#include "nvim/api/private/helpers.h"` header.

@gpanders what do you think?

```suggestion
void ex_trust(exarg_T *eap)
{
  bool deny = false;
  bool remove = false;

  struct {
    String arg;
    bool *flag;
  } flags[] = {
    { STATIC_CSTR_AS_STRING("++deny"), &deny },
    { STATIC_CSTR_AS_STRING("++remove"), &remove },
  };

  const char *p = eap->arg;
  while (p[0] != '\0' && (p[0] == '+' && p[1] == '+')) {
    const char *s = p;
    p = skiptowhite(p);
    size_t len = (size_t)(p - s);

    bool match = false;
    for (size_t i = 0; i < ARRAY_SIZE(flags); i++) {
      String arg = flags[i].arg;
      bool *flag = flags[i].flag;
      if (len == arg.size && strncmp(s, arg.data, arg.size) == 0) {
        if (*flag) {
          semsg(_(e_duparg2), arg.data);
          return;
        }
        *flag = true;
        match = true;
      }
    }

    if (!match) {
      semsg(_(e_invarg2), s);
      return;
    }

    p = skipwhite(p);
  }

  if (deny && remove) {
    emsg(_(e_invarg));  // TODO: a better error message. those are mutually exclusive
    return;
  }

  nlua_trust(deny ? "deny" : remove ? "remove" : "allow", p[0] != '\0' ? p : NULL);
}
```

Reply:
Yea that looks good, though we can generalize that beyond just arg parsing for this function. Let's punt that as a refactor task for a follow-up PR so we can close this one out.

Reply:
Captured here: https://github.com/neovim/neovim/issues/21211